File: 3459-find-the-minimum-area-to-cover-all-ones-ii/find-the-minimum-area-to-cover-all-ones-ii.go
type key struct {
	i1, j1, i2, j2, k int
}

func minimumSum(grid [][]int) int {
	memo := make(map[key]int)
	m, n := len(grid), len(grid[0])

	// Helper: area of bounding rectangle for all 1s in subgrid
	var getOne func(int, int, int, int) int
	getOne = func(i1, j1, i2, j2 int) int {
		minx, maxx := math.MaxInt32, math.MinInt32
		miny, maxy := math.MaxInt32, math.MinInt32

		for i := i1; i <= i2; i++ {
			for j := j1; j <= j2; j++ {
				if grid[i][j] == 1 {
					if i < minx {
						minx = i
					}
					if i > maxx {
						maxx = i
					}
					if j < miny {
						miny = j
					}
					if j > maxy {
						maxy = j
					}
				}
			}
		}
		if minx == math.MaxInt32 {
			return 0
		}
		return (maxx - minx + 1) * (maxy - miny + 1)
	}

	// Recursive with memo
	var getNext func(int, int, int, int, int) int
	getNext = func(i1, j1, i2, j2, k int) int {
		kd := key{i1, j1, i2, j2, k}
		if val, ok := memo[kd]; ok {
			return val
		}

		res := math.MaxInt32
		if k == 1 {
			res = getOne(i1, j1, i2, j2)
		} else if k == 2 {
			// Horizontal splits
			for i := i1; i < i2; i++ {
				a := getNext(i1, j1, i, j2, 1) + getNext(i+1, j1, i2, j2, 1)
				if a < res {
					res = a
				}
			}
			// Vertical splits
			for j := j1; j < j2; j++ {
				a := getNext(i1, j1, i2, j, 1) + getNext(i1, j+1, i2, j2, 1)
				if a < res {
					res = a
				}
			}
		} else if k == 3 {
			// Horizontal splits
			for i := i1; i < i2; i++ {
				a := getNext(i1, j1, i, j2, 1) + getNext(i+1, j1, i2, j2, 2)
				if a < res {
					res = a
				}
				b := getNext(i1, j1, i, j2, 2) + getNext(i+1, j1, i2, j2, 1)
				if b < res {
					res = b
				}
			}
			// Vertical splits
			for j := j1; j < j2; j++ {
				a := getNext(i1, j1, i2, j, 1) + getNext(i1, j+1, i2, j2, 2)
				if a < res {
					res = a
				}
				b := getNext(i1, j1, i2, j, 2) + getNext(i1, j+1, i2, j2, 1)
				if b < res {
					res = b
				}
			}
		}

		memo[kd] = res
		return res
	}

	return getNext(0, 0, m-1, n-1, 3)
}

File: 3459-find-the-minimum-area-to-cover-all-ones-ii/find-the-minimum-area-to-cover-all-ones-ii.py
class Solution:
    def minimumSum(self, grid: List[List[int]]) -> int:
        from collections import defaultdict
        memo = defaultdict(int)  # memoization cache

        # Function: find bounding rectangle area of all 1's inside subgrid
        def get_one(i1, j1, i2, j2):
            minx = float('inf')
            maxx = float('-inf')
            miny = float('inf')
            maxy = float('-inf')

            # Scan subgrid
            for i in range(i1, i2 + 1):
                for j in range(j1, j2 + 1):
                    if grid[i][j] == 1:
                        minx = min(minx, i)
                        maxx = max(maxx, i)
                        miny = min(miny, j)
                        maxy = max(maxy, j)

            # If no '1' inside subgrid
            if minx == float('inf'):
                return 0

            # Return area of bounding rectangle
            return (maxx - minx + 1) * (maxy - miny + 1)

        # Recursive function: minimum sum of areas using k rectangles
        def get_next(i1, j1, i2, j2, k):
            key = (i1, j1, i2, j2, k)
            if key in memo:
                return memo[key]

            output = float('inf')

            if k == 1:  # just one rectangle
                output = get_one(i1, j1, i2, j2)

            elif k == 2:  # split into two rectangles
                # Horizontal splits
                for i in range(i1, i2):
                    output = min(output,
                                 get_next(i1, j1, i, j2, 1) + get_next(i + 1, j1, i2, j2, 1))
                # Vertical splits
                for j in range(j1, j2):
                    output = min(output,
                                 get_next(i1, j1, i2, j, 1) + get_next(i1, j + 1, i2, j2, 1))

            elif k == 3:  # split into three rectangles
                # Horizontal first
                for i in range(i1, i2):
                    # 1 + 2 split
                    output = min(output,
                                 get_next(i1, j1, i, j2, 1) + get_next(i + 1, j1, i2, j2, 2))
                    # 2 + 1 split
                    output = min(output,
                                 get_next(i1, j1, i, j2, 2) + get_next(i + 1, j1, i2, j2, 1))
                # Vertical splits
                for j in range(j1, j2):
                    output = min(output,
                                 get_next(i1, j1, i2, j, 1) + get_next(i1, j + 1, i2, j2, 2))
                    output = min(output,
                                 get_next(i1, j1, i2, j, 2) + get_next(i1, j + 1, i2, j2, 1))

            memo[key] = output
            return output

        m, n = len(grid), len(grid[0])
        # Start with whole grid, need 3 rectangles
        return get_next(0, 0, m - 1, n - 1, 3)
